### PR TITLE
[issue 51] Add slash command autocomplete functionality to Bitbucket #51

### LINKED ISSUE #51: Add slash command autocomplete functionality to Bitbucket
Bitbucket does not have any auto-complete functionality for the available subcommands.

Given how many subcommands are available this would be very useful. 

Steps:
- Type `/bitbucket` and a space. 

**Observed**No subcommands are shown.

### PR DESCRIPTION
**Pull request by sibasankarnayak**
added slash command for plugin

ticket here
Fixes #51

### CODE REVIEW

File: server/command.go
Comment (mickmister) [inline server/command.go]:
```suggestion
	connect := model.NewAutocompleteData("connect", "", "Connect your Mattermost account to your Bitbucket account")
```

File: server/command.go
Comment (mickmister) [inline server/command.go]:
```suggestion
	disconnect := model.NewAutocompleteData("disconnect", "", "Disconnect your Mattermost account from your Bitbucket account")
```

File: server/command.go
Comment (mickmister) [inline server/command.go]:
```suggestion
	subscribeList := model.NewAutocompleteData("list", "", "List subscriptions")
```

Are the subscriptions scoped to the current channel? We should clarify that in the autocomplete suggestion

File: server/command.go
Comment (mickmister) [inline server/command.go]: This looks a little confusing. Do you mind creating a gif showing this command in action?

File: server/command.go
Comment (mickmister) [inline server/command.go]: Why are we adding two subscribe commands?

File: server/command.go
Comment (mickmister) [inline server/command.go]:
```suggestion
	unsubscribe := model.NewAutocompleteData("unsubscribe", "[unsubscribe] [owner/repo]", "Remove subscription for org/[repo]")
```

**Review by mickmister — CHANGES_REQUESTED** [review]
Mostly looking good, just have a few requests. Thanks @sibasankarnayak!

File: server/command.go
Comment (sibasankarnayak) [reply server/command.go]: @mickmister 
the current model of adding a subscription is in such a way 

```
/bitbucket subscribe list 
/bitbucket subscribe [owner/repo]  [features]
```

was not sure hoe to add NewAutocompleteData , AddTextArgument in same level so tried in this way.

File: server/command.go
Comment (sibasankarnayak) [reply server/command.go]: @mickmister sure

https://user-images.githubusercontent.com/86650698/150576340-75645f0e-c5a1-4bfb-be41-65b6aa10cb7d.mov

File: server/command.go
Comment (hanzei) [reply server/command.go]: @sibasankarnayak Did you make these changes?

File: server/command.go
Comment (hanzei) [reply server/command.go]: @sibasankarnayak Did you make these changes?

File: server/command.go
Comment (hanzei) [inline server/command.go]: Should the slash command work similar as in GitHub and be `subscriptions`? Then this would become

```suggestion
	subscriptions := model.NewAutocompleteData("subscriptions", "[command]", "Available commands: list, add")
```

Surely, the sub commands need to be adjusted then.

**Comment by sibasankarnayak** [comment]
@mickmister @hanzei  added this two feature in this PR too https://github.com/mattermost/mattermost-plugin-github/issues/513 https://github.com/mattermost/mattermost-plugin-github/issues/512

File: server/command.go
Comment (hanzei) [inline server/command.go]:
```suggestion
* |/bitbucket subscriptions add owner [features]| - Subscribe the current channel to all available repositories within an organization and receive notifications about opened pull requests and issues
```

File: server/command.go
Comment (hanzei) [inline server/command.go]:
```suggestion
* |/bitbucket subscriptions add owner/repo [features]| - Subscribe the current channel to receive notifications about opened pull requests and issues for a repository
```

File: server/command.go
Comment (hanzei) [inline server/command.go]: `unsubscribe` should also be a subcommand of `subscriptions`. I would suggest using the same handler as in https://github.com/mattermost/mattermost-plugin-github/blob/4db6b6a461d7d1ec5e137ee8a744b6ced3ed4e2d/server/plugin/command.go#L224-L242

**Review by hanzei — CHANGES_REQUESTED** [review]

**Review by hanzei — APPROVED** [review]
I still think the code in https://github.com/mattermost/mattermost-plugin-github/blob/4db6b6a461d7d1ec5e137ee8a744b6ced3ed4e2d/server/plugin/command.go#L224-L242 is cleaner, but this is also fine.

**Comment by sibasankarnayak** [comment]
> I still think the code in https://github.com/mattermost/mattermost-plugin-github/blob/4db6b6a461d7d1ec5e137ee8a744b6ced3ed4e2d/server/plugin/command.go#L224-L242 is cleaner, but this is also fine.

@hanzei should i make it clean ?

**Comment by hanzei** [comment]
Your call. I'm fine with the current code.

**Comment by sibasankarnayak** [comment]
> Your call. I'm fine with the current code.

would perfer to keep it

File: server/command.go
Comment (mickmister) [inline server/command.go]: Are these per channel?

```suggestion
	subscribeList := model.NewAutocompleteData("list", "", "List Subscription for this channel")
```

File: server/command.go
Comment (mickmister) [reply server/command.go]: @dipak-demansol Can you please make sure the autocomplete works correctly in both cases of @sibasankarnayak's example above?

File: server/command.go
Comment (mickmister) [inline server/command.go]: 0/5 for all references to the generic `subscriptions` command, rather than referring to the act of subscribing in some cases

```suggestion
	subscriptions := model.NewAutocompleteData("subscriptions", "[command]", "Available commands: list, add")
	subscriptionsList := model.NewAutocompleteData("list", "", "List Subscription")
	subscriptions.AddCommand(subscriptionsList)
```

**Review by mickmister — APPROVED** [review]
Awesome work @sibasankarnayak, this makes the plugin a lot more user-friendly. Approving with some suggestions

**Comment by hanzei** [comment]
@sibasankarnayak Please take a look at @mickmister and let me know what you think

**Comment by dipak-demansol** [comment]
@sibasankarnayak pls ping me when everything is done from your side so i can test it.

File: server/command.go
Comment (sibasankarnayak) [reply server/command.go]: @mickmister 

current command is something like this
```
/bitbucket subscriptions list 
/bitbucket subscriptions add  owner[/repo]
/bitbucket subscriptions delete owner[/repo]

```

cc @dipak-demansol

**Comment by sibasankarnayak** [comment]
> @sibasankarnayak pls ping me when everything is done from your side so i can test it.

@dipak-demansol Please test this and share your feedback

**Comment by hanzei** [comment]
@dipak-demansol Gentle reminder to review this PR

**Comment by dipak-demansol** [comment]
> @dipak-demansol Gentle reminder to review this PR

Testing is done, There are few point that i need to discuss with dylan so end of the day i'll add my comment.

**Review by dipak-demansol — CHANGES_REQUESTED** [review]
@hanzei @mickmister @sibasankarnayak 

Non-Blocking Point
---------------------------
1) when i type "**/bitbucket S**" then show "subscribe command on top, currently its below the settings", why :- subscription will be the most used command compare to setting so its best to display it on the top, same as github, i know that, by type "/bitbucket su" will give me the exact subscription command still it would be great if you show it on The top just by typing "/bitbucket s" 

Blocking Point
----------------------
1) when i Use "**/bitbucket subscriptions add** " then it should give me validation error msg as **"Please specify a org/[repo]"** same as github but Currently its subscribing the "add", same issue when i use "**/bitbucket subscriptions delete**" & "**/bitbucket subscriptions list**" Command, pls See The Attachment.
image

**Comment by catalintomai** [comment]
@sibasankarnayak , can you address @dipak-demansol 's comments? We were targeting the release of this plugin for last month. Thanks.

**Comment by sibasankarnayak** [comment]
> @sibasankarnayak , can you address @dipak-demansol 's comments? We were targeting the release of this plugin for last month. Thanks.

@catalintomai sure
@dipak-demansol can you check now.

**Comment by dipak-demansol** [comment]
> > @sibasankarnayak , can you address @dipak-demansol 's comments? We were targeting the release of this plugin for last month. Thanks.
> 
> @catalintomai sure @dipak-demansol can you check now.

@sibasankarnayak 
1) When i Use "//bitbucket  subscriptions" AND add any parameter ex  "/bitbucket subscriptions ABCD" then its add "ABCD" but i did not used the "add" Command, The user should get the validation message.
image

2) With the list Command if i pass any extra parameter then its behaving as "add" Functionality.
image

**Comment by sibasankarnayak** [comment]
> 1. /bitbucket subscriptions ABCD

@dipak-demansol thank you for finding such cases,
 can you test now

**Review by dipak-demansol — APPROVED** [review]
LGTM

**Pull request merged**
